### DISCUSSION TITLE
deleting a lot of files takes a long time on macOS

### DISCUSSION THREAD

**Discussion by verhovsky** [magit/magit#4635] [Support] [answered]
To reproduce, on macOS, `cd` into a git repo and create 700 files

```
touch {1..700}.txt
```

open the repo with magit then select all 700 untracked files and press x to trash them and press y to confirm.

#### what I expected:

All files are deleted in one go

#### what I ~~observed~~ heard:

machine gun fire as my macbook rapidly played the move to trash sound 700 times for like 20 seconds.

Edit: it's actually closer to 2 minutes I think. And if you're curious about my use case, I have a script that generates like 900 test files for all possible combinations of some values. I have about 200 test files that exist of those 900, so it's a way of quickly updating a bunch of autogenerated test files basically, and it's faster (it would be faster) for me to run the script and unstage all the extra files it creates than to add code to the script to make it only output files that already exist, and I wouldn't necessarily want that because sometimes I want to create new test files.

**Comment by tarsius** [comment]
> select all 700 untracked files and press x to trash them

`x`? Are you using `evil`?

Instead move point onto the heading "Untracked files" and then press `k` (`magit-discard`).

By the way, for me selecting files and then pressing `k` (`magit-discard`) also deletes the files in less than a second. Maybe Magit isn't compiled for you?

Or maybe this is some macOS "access control" feature kicking in.

**Reply by verhovsky** [reply]
> x? Are you using evil?

yes, through Doom Emacs.

> move point onto the heading "Untracked files" and then press k

same thing

> some macOS "access control" feature kicking in

I think it's just taking so long because it has to play the trash sound 700 times and it takes like 200 ms to play it or whatever.

**Reply by tarsius** [reply]
What's the "it" that has to play the trash sound? Emacs or macOS?

Magit uses `dired-delete-file`, which then uses `delete-file`. I took a quick look and didn't see anything there that might play a sound.

**Reply by verhovsky** [reply]
macOS, more specifically Finder.app I think? 

If it helps, there's a 3rd-party `trash` command (https://github.com/ali-rantakari/trash or `brew install trash`)  that has this flag

```
  -F  Ask Finder to move the files to the trash, instead of
      using the system API.
```

and `trash -F some_file.txt` plays the noise from my terminal.

I think somewhere in the emacs code there's a for loop, when it should pass a list of files to wherever it's telling Finder to delete the files.

**Reply by tarsius** [reply]
> I think somewhere in the emacs code there's a for loop, when it should pass a list of files to wherever it's telling Finder to delete the files.

Yes. It is spread across `magit-discard-files` and `magit-discard-files--delete`. Also see the surrounding complexity. The problem essentially is that then "discarding" multiple files the individual files might have to be handled quite differently, even though they appear in the same list of files. This is already complex and I do not wish to add a special case for the "delete several untracked" files.

IMO macOS should not play a sound when a third-party application deletes a file. It's perfectly find if it does that in the Finder or if a third-party app has opted in to fancy behavior, but to do this at such a low level as it apparently does, is a mistake. And it is a mistake that I am not willing to pay the price for; I am not going to complicate my code to undo their mistake. Sorry.

**Reply by rudolf-adamkovic** [reply]
_Note:_ macOS does not play any sounds. See my comment (re-test) below.

**Reply by verhovsky** [reply]
@rudolf-adamkovic you need to set

```emacs
(setq delete-by-moving-to-trash t)
```

and also

```sh
brew install trash
```

But macOS 14 (released 2023) added a native `trash` command that doesn't play the sound.

**Comment by rudolf-adamkovic** [comment]
I followed the reproduction steps on my machine and Magit deleted the 700 files before I started my stopwatch.  The machine played no sounds.  (I have never install any third-party "trash" programs.) macOS 12.1 (21C52), Magit 20220312.452, Git 2.35.1, Emacs 29.0.50. (compiled ~2 days ago).

**Reply by tarsius** [reply]
And was it slow or fast?

**Reply by rudolf-adamkovic** [reply]
Under a second.  I wanted to use my Casio watch to measure.  Way too fast.  :)

**Comment by tarsius** [comment, marked as answer]
Emacs does not provide a `delete-files` (plural) function.

In this situation `dired-delete-file` is basically the same as `delete-file` (plus the cost of calling `file-attributes`).

I am guessing the problem is that by default files are moved to the trash and that the macOS trash is slow. Try `(setq magit-delete-by-moving-to-trash nil)`.

**Reply by verhovsky** [reply]
that works great, thanks!
